### PR TITLE
feat: mainnet snapshot and local node deployment from it

### PR DESCRIPTION
**Pull request by kupermind**
- Ability to do a full snapshot of a current registries state;
- The local deployment script now requires a `snapshot.json` taken in the form from one of previous snapshots to deploy the same set of components / agents / services

### CODE REVIEW

File: deploy/contracts.js [hunk 1]
```diff
@@ -46,30 +37,34 @@ module.exports = async () => {
...
+        await componentRegistry.connect(deployer).create(deployer.address,
+            snapshotJSON["componentRegistry"]["hashes"][i], snapshotJSON["componentRegistry"]["dependencies"][i]);
+    }
+
+    // Create agents from the snapshot data
+    const numAgents = snapshotJSON["agentRegistry"]["hashes"].length;
+    for (let i = 0; i < numAgents; i++) {
+        await agentRegistry.connect(deployer).create(deployer.address,
+            snapshotJSON["agentRegistry"]["hashes"][i], snapshotJSON["agentRegistry"]["dependencies"][i]);
+    }
```
Comment (kupermind) [inline deploy/contracts.js:61]: Constructed from provided snapshot JSON file

File: deploy/contracts.js [hunk 2]
```diff
@@ -85,83 +80,115 @@ module.exports = async () => {
...
+    // For simplicity, deployer is the manager for service registry
+    await serviceRegistry.changeManager(deployer.address);
+
+    // Create and deploy services based on the snapshot
+    const numServices = snapshotJSON["serviceRegistry"]["configHashes"].length;
+    // Agent instances cannot repeat, so each of them must be a unique address
+    // TODO: With a bigger number of services, precalculate the number of agent instnaces first and allocate enough addresses for that
+    let aCounter = 0;
+    console.log("==========================================");
+    for (let i = 0; i < numServices; i++) {
```
Comment (kupermind) [inline deploy/contracts.js:119]: Constructed from provided snapshot JSON file

File: scripts/mainnet_snapshot.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+async function main() {
+    // This script is designed specifically for the mainnet deployment
+    // Getting the registry contracts
+    const componentRegistry = await ethers.getContractAt("ComponentRegistry", "0x15bd56669F57192a97dF41A2aa8f4403e9491776");
+    const agentRegistry = await ethers.getContractAt("AgentRegistry", "0x2F1f7D38e4772884b88f3eCd8B6b9faCdC319112");
+    const serviceRegistry = await ethers.getContractAt("ServiceRegistry", "0x48b6af7B12C71f09e2fC8aF4855De4Ff54e775cA");
+
+    // Initialize the snapshot structure
+    let snapshotJSON = {
```
Comment (kupermind) [inline scripts/mainnet_snapshot.js:13]: The snapshot JSON structure

File: scripts/mainnet_snapshot.js [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        snapshotJSON["componentRegistry"]["hashes"].push(unit.unitHash);
+        //console.log("component hash", unit.unitHash);
+    }
+
+    // Get the agent data
+    for (let i = 1; i <= numAgents; i++) {
+        const unit = await agentRegistry.getUnit(i);
+        snapshotJSON["agentRegistry"]["dependencies"].push(unit.dependencies);
+        snapshotJSON["agentRegistry"]["hashes"].push(unit.unitHash);
+    }
```
Comment (kupermind) [inline scripts/mainnet_snapshot.js:56]: Getting components and agents data

File: scripts/mainnet_snapshot.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    // Get the service data
+    for (let i = 1; i <= numServices; i++) {
```
Comment (kupermind) [inline scripts/mainnet_snapshot.js:59]: Getting services data

File: deploy/contracts.js [hunk 6]
```diff
@@ -30,6 +20,7 @@ module.exports = async () => {
         "0x5de4111afa1a4b94908f83103eb1f1706367c2e68ca870fc3fb9a804cdab365a",
         "0x7c852118294e51e653712a81e05800f419141751be58f605c371e15141b007a6",
     ];
+    const operatorPK = "0xf214f2b2cd398c806f84e317254e0f0b801d0643303237d97a22a48e01628897";
```
Comment (DavidMinarsch) [inline deploy/contracts.js:23]: ?

File: deploy/contracts.js [hunk 7]
```diff
@@ -85,83 +80,115 @@ module.exports = async () => {
...
-        if (!isOwner) {
-            throw new Error("incorrect agent instance");
+
+    // For simplicity, deployer is the manager for service registry
+    await serviceRegistry.changeManager(deployer.address);
+
+    // Create and deploy services based on the snapshot
+    const numServices = snapshotJSON["serviceRegistry"]["configHashes"].length;
+    // Agent instances cannot repeat, so each of them must be a unique address
+    // TODO: With a bigger number of services, precalculate the number of agent instnaces first and allocate enough addresses for that
```
Comment (DavidMinarsch) [inline deploy/contracts.js:116]: typo `instnaces`

File: deploy/contracts.js (hunk 6 above)
Comment (kupermind) [reply deploy/contracts.js:23]: That's a default operator private key take from the default hardhat signers[10], that might be required at some point

File: README.md [hunk 8]
```diff
@@ -105,6 +105,14 @@ The audit is provided as development matures. The latest audit report can be fou
 ## Deployed Protocol
 The list of addresses can be found [here](https://github.com/valory-xyz/autonolas-registries/blob/main/docs/mainnet_addresses.json).
 
+### Mainnet snapshot of registries
+In order to get the current snapshot of all the registries, the following script is provided [here](https://github.com/valory-xyz/autonolas-registries/blob/main/scripts/mainnet_snapshot.js).
+The script can be run with the following command:
+```
+npx hardhat run scripts/mainnet_snapshot.js --network mainnet
+```
+Please note that for the correct mainnet interaction the `ALCHEMY_API_KEY` needs to be exported as an environment variable.
```
Comment (DavidMinarsch) [inline README.md:114]:
```suggestion
Please note that for the correct mainnet interaction the `ALCHEMY_API_KEY` needs to be exported as an environment variable.

NOTE: whilst the snapshot does maintain the exact dependency structure between components, agents and services, it does not conserve the ownership structure.
```

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**